lib/app-infos-cache.js

```diff
@@ -22,130 +25,134 @@ export class AppInfosCache {
+  async extractBundleVersion (bundlePath) {
+    return await this.extractManifestProperty(bundlePath, 'CFBundleVersion');
   }
 
   /**
    *
-   * @param {string} appPath
+   * @param {string} bundlePath Full path to the .ipa or .app bundle
    * @returns {Promise<string[]>}
    */
-  async extractAppPlatforms (appPath) {
-    const result = await this.extractManifestProperty(appPath, 'CFBundleSupportedPlatforms');
+  async extractAppPlatforms (bundlePath) {
+    const result = await this.extractManifestProperty(bundlePath, 'CFBundleSupportedPlatforms');
     if (!Array.isArray(result)) {
-      throw new Error(`${path.basename(appPath)}': CFBundleSupportedPlatforms is not a valid list`);
+      throw new Error(`${path.basename(bundlePath)}': CFBundleSupportedPlatforms is not a valid list`);
     }
     return result;
   }
 
   /**
    *
-   * @param {string} appPath
+   * @param {string} bundlePath Full path to the .ipa or .app bundle
    * @returns {Promise<string>}
    */
-  async extractExecutableName (appPath) {
-    return await this.extractManifestProperty(appPath, 'CFBundleExecutable');
+  async extractExecutableName (bundlePath) {
+    return await this.extractManifestProperty(bundlePath, 'CFBundleExecutable');
   }
 
   /**
    *
-   * @param {string} appPath Full path to the .ipa or .app bundle
+   * @param {string} bundlePath Full path to the .ipa or .app bundle
    * @returns {Promise<import('@appium/types').StringRecord>} The payload of the manifest plist
    * @throws {Error} If the given app is not a valid bundle
    */
-  async put (appPath) {
-    const readPlist = async (/** @type {string} */ plistPath) => {
-      try {
-        return await plist.parsePlistFile(plistPath);
-      } catch (e) {
-        this.log.debug(e.stack);
-        throw new Error(`Cannot parse ${MANIFEST_FILE_NAME} of '${appPath}'. Is it a valid application bundle?`);
-      }
-    };
+  async put (bundlePath) {
+    return (await fs.stat(bundlePath)).isFile()
+      ? await this._putIpa(bundlePath)
+      : await this._putApp(bundlePath);
+  }
 
-    if ((await fs.stat(appPath)).isFile()) {
-      /** @type {import('@appium/types').StringRecord|undefined} */
-      let manifestPayload;
-      /** @type {Error|undefined} */
-      let lastError;
-      try {
-        await zip.readEntries(appPath, async ({entry, extractEntryTo}) => {
-          if (!_.endsWith(entry.fileName, `.app/${MANIFEST_FILE_NAME}`)) {
-            return true;
-          }
+  /**
+   * @param {string} ipaPath Fill path to the .ipa bundle
+   * @returns {Promise<import('@appium/types').StringRecord>} The payload of the manifest plist
+   */
+  async _putIpa(ipaPath) {
+    /** @type {import('@appium/types').StringRecord|undefined} */
+    let manifestPayload;
+    /** @type {Error|undefined} */
+    let lastError;
+    try {
+      await zip.readEntries(ipaPath, async ({entry, extractEntryTo}) => {
+        if (!IPA_ROOT_PLIST_PATH_PATTERN.test(entry.fileName)) {
```

Comment:
so, the main change is here?
The top of info plist (Payload/xxx.app/Info.plist) will be read. Watch app would be `Payload/xxx.app/Watch/yyy.app` so watch app will be ignored.

Can you leave a comment? I may forget the nested Watch app case here in the future...

(just a link I checked the structure)
https://medium.com/@programmingpassion/watchos-app-bundle-release-management-b6e58545b1c9

Reply:
Added the link to the original issue